tools/enrollment_preprocessor.py

```diff
@@ -0,0 +1,86 @@
+#!/usr/bin/env python3
+import csv
+import sys
+from argparse import ArgumentParser
+from dataclasses import dataclass
+from io import BytesIO
+from typing import TextIO
+
+from openpyxl import load_workbook
+from openpyxl.cell import Cell
+
+
+@dataclass
+class User:
+    title: str
+    last_name: str
+    first_name: str
+    email: str
+
+
+@dataclass
+class UserCells:
+    title: Cell | None
+    last_name: Cell
+    first_name: Cell
+    email: Cell
+
+    def value(self) -> User:
+        return User(
+            str(self.title.value) or "" if self.title else "",
+            str(self.last_name.value),
+            str(self.first_name.value),
+            str(self.email.value),
+        )
+
+
+def user_decision(field: str, existing: str, imported: str) -> str:
+    if existing == imported:
+        return existing
+    decision = input(f"Do you want to keep the existing user {field}? (y/n) ")
+    if decision and decision[0].lower() == "n":
+        return imported
+    return existing
+
+
+def fix_user(users: dict[str, User], imported_cells: UserCells) -> None:
+    imported = imported_cells.value()
+    existing = users.setdefault(imported.email, imported)
+    if imported == existing:
+        return
+    print("There is a conflict in the user data.")
+    print(f"existing: {existing}.")
+    print(f"imported: {imported}.")
+    if imported_cells.title is not None:
+        imported_cells.title.value = user_decision("title", existing.title, imported.title)
+    imported_cells.last_name.value = user_decision("last name", existing.last_name, imported.last_name)
+    imported_cells.first_name.value = user_decision("first name", existing.first_name, imported.first_name)
+    imported_cells.email.value = user_decision("email", existing.email, imported.email)
+    print()
+
+
+def run_preprocessor(enrollment_data: str | BytesIO, user_data: TextIO) -> None:
+    workbook = load_workbook(enrollment_data)
+    users = {}
+    reader = csv.reader(user_data, delimiter=";", lineterminator="\n")
+    for row in reader:
+        email = row[-1]
+        users[email] = User(*row)
+    for sheet in workbook.worksheets:
+        for wb_row in sheet.iter_rows(min_row=2, min_col=2):
+            fix_user(users, UserCells(None, *wb_row[:3]))
+            fix_user(users, UserCells(*wb_row[7:]))
+    workbook.save(enrollment_data)
+
+
+if __name__ == "__main__":  # pragma: nocover
+    parser = ArgumentParser(description="Commandline tool to preprocess enrollment xlsx files.")
+    parser.add_argument(
+        "-u", "--user-data", help="Path to a csv file containing an export of all existing users.", required=True
+    )
+    parser.add_argument(
+        "-e", "--enrollment-data", help="Path to the enrollment data in xlsx format for import.", required=True
+    )
```

Comment:
This makes it so that a user _has_ to at least specify `-e` and `-u`, you can't just paste two paths as arguments, right?

If we wanted to keep it simple, we could just make them both positional. I'd say this would be common practice if you have two required arguments that you can't really accidentally switch up (as a .csv file will not pass xlsx loading). But I'm also fine with the current named approach.

Reply:
I added it in [ce8c35e](https://github.com/e-valuation/EvaP/pull/2011/commits/ce8c35e12081ea3211987938c34944d691dd239d), we can revert it, if needed